### PR TITLE
[FLINK-8175] [DataStream API java/scala] remove flink-streaming-contrib and migrate its classes to flink-streaming-java/scala

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

This is the first step to clean up flink-contrib. More details in (FLINK-8188)[https://issues.apache.org/jira/browse/FLINK-8188]

This PR removes `flink-streaming-contrib` from `flink-contrib`, and migrates its classes to `flink-streaming-java/scala` for the following reasons:

- `flink-streaming-contrib` is so small that it only has 4 classes (3 java and 1 scala), and they don't need a dedicated jar for Flink to distribute and maintain it and for users to deal with the overhead of dependency management
- the 4 classes in `flink-streaming-contrib` are logically more tied to `flink-streaming-java/scala`, and thus can be easily migrated
- `flink-contrib` is already too crowded and noisy. It contains lots of sub modules with different purposes which confuse developers and users, and they lack a proper project hierarchy

## Brief change log

This PR removes `flink-streaming-contrib` from `flink-contrib`, and migrates its classes to `flink-streaming-java/scala`

## Verifying this change

This change is already covered by existing tests:

- `SocketStreamIteratorTest` is moved to `flink-streaming-java`
- `CollectITCase` is moved to `flink-tests`

## Does this pull request potentially affect one of the following parts:

none

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs / JavaDocs)

### CODE REVIEW

**Comment by bowenli86** [comment]
Hi @tillrohrmann , per our discussion, I sent out an email to dev@flink, but seems it didn't catch any attention. What do you think about this PR?

**Comment by tillrohrmann** [comment]
Thanks for your contribution @bowenli86. I think we should reach first some kind of consensus on the ML before moving the classes. I'm actually not sure whether `CollectSink` is something we should keep around since it is not at all generic and not efficient enough to be seriously used. My gut feeling is that this is actually a toy sink which should not be part of the `flink-streaming` modules.

**Comment by bowenli86** [comment]
@tillrohrmann  I agree that `CollectSink` seems to be immature after taking another detailed look at it.

I've sent an email to dev alias on Dec 2. Can you please respond to it and start a discussion?

**Comment by StephanEwen** [comment]
I would actually be in favor of dismantling the `flink-streaming-contrib` project.

We could have a dedicated package `experimental` in `flink-streaming-java` for such code. The package should have a `package-info.java` clearly stating that this code is not battle hardened code and may change in future versions. None of the classes should be `@Public`, might could be be `@PublicEvolving`.

**Comment by bowenli86** [comment]
@StephanEwen That's a great idea to have an `/experimental` package for such things. Will update the PR

**Comment by bowenli86** [comment]
@StephanEwen let me know if you have more concerns

**Comment by bowenli86** [comment]
@StephanEwen @tillrohrmann  let me know if you have more concerns

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/experimental/package.scala [hunk 1]
```diff
@@ -16,23 +16,25 @@
  * limitations under the License.
  */
 
-package org.apache.flink.contrib.streaming.scala
+package org.apache.flink.streaming.experimental
```
Comment (StephanEwen) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/experimental/package.scala]: To retain symmetry with the original code, would be good to have this in a `scala` subpackage.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/experimental/CollectITCase.java [hunk 2]
```diff
@@ -7,7 +7,7 @@
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ * http://www.apache.org/licenses/LICENSE-2.0
```
Comment (StephanEwen) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/experimental/CollectITCase.java]: There are a bunch of these types of formatting changes. They actually make the license header different from the other files. Would be good to revert this.

**Comment by bowenli86** [comment]
@StephanEwen Thanks Stephan. To your comments:

1. in order to have `experiemental/scala` package in `flink-streaming-scala`, I have to make `DataStreamUtils` its own class to avoid package name collisions with `scala`
2. I don't think checkstyle of the whole project`flink-streaming-scala` is active. I did some tests which removed comments and left unused class imports, the build still passed. I think we should address enabling checkstyle separately, may need to reopen[FLINK-6902]( https://issues.apache.org/jira/browse/FLINK-6902). cc @zentol

**Comment by greghogan** [comment]
@bowenli86 Checkstyle is Java-only. There is a [Scalastyle](http://www.scalastyle.org).

**Comment by bowenli86** [comment]
@StephanEwen @greghogan ok, so looks like the rules of checkstype and scalastyle are a bit different, that's why I misreported `flink-streaming-scala`'s style checking is inactive.

scala style for subpackage `/experimental.scala` is automatically active already. I did some tests and made sure that.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/experimental/CollectSink.java [hunk 3]
```diff
@@ -28,8 +29,12 @@
 import java.net.Socket;
 
 /**
- * A specialized data sink to be used by DataStreamUtils.collect.
+ * A specialized data sink to be used by DataStreamUtils.collect().
+ *
+ * <p>This experimental class is relocated from flink-streaming-contrib. Please see package-info.java
+ * for more information.
  */
+@PublicEvolving
```
Comment (StephanEwen) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/experimental/CollectSink.java]: Would this better be `@Internal`? I think it is not meant to be instantiated directly, but only via the collect() call...

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/experimental/package-info.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This package holds classes that are experimental.
+ *
+ * They are NOT battle-tested code and may be changed or removed in future versions.
```
Comment (StephanEwen) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/experimental/package-info.java]: For checkstyle, you probably need to prepend new paragraphs with `<p>`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/experimental/CollectSink.java (hunk 3 above)
Comment (bowenli86) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/experimental/CollectSink.java]: make sense. Changed

**Comment by bowenli86** [comment]
@StephanEwen  let me know if this is good to go

**Pull request closed without merging**